docs/document location prop (#17918)
gatsbyjs/gatsby: docs/docs/location-data-from-props.md

---
title: "Location Data from Props"
---

## What is location data

> Sometimes it can be helpful to know exactly what your app's browser url is at any given stage. Using @reach/router, location data represents where the app is currently, where you'd like it to go, and other helpful information. The location object is never mutated but `reach@router` this makes it helpful to determine when navigation happens.

Author: I agree. I think there is an opportunity with the changes / merge of react-router + @reach/router to really bolster the tutorials, education, and documentation around this feature.

> ```js:title=location object

Author: much better without the title 😁

{

>   key: 'ac3df4', // does not populate with a HashHistory!

Author: Added a small section going over `HashHistory` and a link in the resources that provides a more in depth explanation.

  pathname: '/somepage',
  search: '?someurlquery=searching-string',
  hash: '#about',
  state: {
    [userDefined]: true
  }
}
```

## Use cases

> You can provide a location object instead of strings, helpful in a number of situations.

Author: Something that would show what is auto-populated by the router versus what you can set in the location object itself. Is that sort of what we're looking for? For example search created from the browser URL and can't be set programmatic that I know of, but pathname and state are properties are usually set on location object at the component site, and not by the router.

- Providing state to linked components
- Client-only routes
- Fetching Data
- Animation Transition

## Example

```js:title=index.js
// usually you'd do this
<Link to="/somecomponent"/>

// but if we want to add some additional state
<Link to={
  pathname: '/somecomponent',
  state: {modal: true}
}>
```

> Then from the recieving component we can conditionally render it based on the location state.

Author: will get used to the 'you' tone from the style guide. 👍

```js:title=some-component.js
const SomeComponent = ({ location }) => {
  const { state = {} } = location
  const { modal } = state
  return modal ? (
    <div className="modal">I'm a modal of Some Component!</div>
  ) : (
    <div>Welcome to the Some Component page!</div>
  )
}
```

## Gatsby advantages

The great thing is you can expect the location prop to be available to you on everypage thanks to its use of `@reach/router`

## Other resources

- [@reach/router docs](https://reach.tech/router/api/Location)
- [react-router location docs](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/location.md)

> - [Gatsby Breadcrumb Plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-breadcrumb/#breadcrumb-props)

Reviewer: ```suggestion
- [Gatsby Breadcrumb Plugin](/packages/gatsby-plugin-breadcrumb/#breadcrumb-props)
```
We can use relative links for gatsbyjs.org!

Author: that's awesome! 🎉

---
title: "Location Data from Props"
---

## What is location data

> Sometimes it can be helpful to know exactly what your app's browser URL is at any given stage. Using [@reach/router](https://github.com/reach/router) for [client-side routing](/docs/glossary#client-side) in Gatsby, `location` data represents where the app is currently, where you'd like it to go, and other helpful information. The `location` object is never mutated but `reach@router` makes it helpful to determine when navigation happens.

Reviewer: ```suggestion
Sometimes it can be helpful to know exactly what your app's browser URL is at any given stage. Using [@reach/router](https://github.com/reach/router) for [client-side](/docs/glossary#client-side) routing in Gatsby, `location` data represents where the app is currently, where you'd like it to go, and other helpful information. The `location` object is never mutated but `reach@router` makes it helpful to determine when navigation happens.
```
That glossary entry doesn't mention routing, so adjusting the link will make it more relevant.

```js
{
  key: 'ac3df4', // does not populate with a HashHistory!
  pathname: '/somepage',
  search: '?someurlquery=searching-string',
  hash: '#about',
  state: {
    [userDefined]: true
  }
}
```

> ## HashHistory

Reviewer: ```suggestion
### HashHistory
```
This is sort-of an extra tidbit about what location data is, so it makes sense to be structured under that heading in the page hierarchy.

> Using `hash` is one way to update the browser URL and the DOM without having the browser do a full reload. The HashHistory is used to track browser history when using [hashrouter](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/HashRouter.md#hashrouter) instead of [browserrouter](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/BrowserRouter.md#browserrouter) which uses the newer HTML5 history API.

Reviewer: ```suggestion
Using `hash` in JavaScript is one way to update the browser URL and the DOM without having the browser do a full HTML page reload. HashHistory in `@reach/router` is used to track browser history with JavaScript when using [hashrouter](https://reacttraining.com/react-router/web/api/HashRouter) instead of [browserrouter](https://reacttraining.com/react-router/web/api/BrowserRouter) which uses the newer HTML5 `history` API.
```
Updating to the website front-end of their API docs and adding a little more context.

## Use cases

> Through client-side routing in Gatsby you can provide a location object instead of strings, helpful in a number of situations.

Reviewer: ```suggestion
Through client-side routing in Gatsby you can provide a location object instead of strings, which are helpful in a number of situations:
```

- Providing state to linked components
- Client-only routes
- Fetching Data
- Animation Transition

> ## Example

Reviewer: ```suggestion
## Example of providing state to a link component
```

```js:title=index.js
// usually you'd do this

> <Link to="/somecomponent"/>

Reviewer: ```suggestion
<Link to="/somepagecomponent"/>
```
To make it clear that links go to pages, mentioning `page` in the component name seems more intuitive.

// but if you want to add some additional state
<Link to={

>   pathname: '/somecomponent',

Reviewer: ```suggestion
  pathname: '/somepagecomponent',
```

  state: {modal: true}
}>
```

Then from the receiving component you can conditionally render markup based on the `location` state.

> ```js:title=some-component.js

Reviewer: ```suggestion
```js:title=some-page-component.js
```

> const SomeComponent = ({ location }) => {

Reviewer: ```suggestion
const SomePageComponent = ({ location }) => {
```

  const { state = {} } = location
  const { modal } = state
  return modal ? (

>     <div className="modal">I'm a modal of Some Component!</div>

Reviewer: ```suggestion
    <dialog className="modal">I'm a modal of Some Page Component!</dialog>
```

  ) : (

>     <div>Welcome to the Some Component page!</div>

Reviewer: ```suggestion
    <div>Welcome to the Some Page Component page!</div>
```

Author: Appreciate the added context of 'Page' to the example. It's subtle, but definitely more helpful as a resource.
